Implement PPO-DNA algorithm for Atari

jseppanen opened the pull request:
## Description

Add implementation of PPO-DNA algorithm for Atari Envpool.

### Paper reproduction (attempt)

Here's the episodic rewards after 200M environment steps (50M gradient updates), compared to Fig. 6 in the original paper:
* BattleZone: 121 000 ± 26 000 (better than about 60 000 in the paper)
* DoubleDunk: -3.8 ± 0.4 (worse than about 1.0 in the paper)
* NameThisGame: 10 700 ± 4 300 (worse than about 20 000 in the paper)
* Phoenix: 208 000 ± 73 000 (better than about 80 000 in the paper)
* Qbert: 12 000 ± 5 000 (worse than about 30 000 in the paper)

However, I used the default networks and environments from the CleanRL PPO Atari implementation, so there are probably differences in them, vs. the original paper. In summary, when comparing against paper results, this implementation gets better returns in two out of the five tasks.

Results from figure 6 in the paper:
<img height="400" alt="figure6" src="https://user-images.githubusercontent.com/83203/179800127-5ccfb2bc-87be-488c-bd7b-8f4c0e8f8103.png">

When comparing against ClearRL PPO Atari Envpool implementation, this implementation performs better on six out of nine tasks. See detailed learning curves, compared against CleanRL PPO Atari Envpool implementation:
[PPO-DNA vs PPO on Atari Envpool](https://wandb.ai/jseppanen/cleanrl/reports/PPO-DNA-vs-PPO-on-Atari-Envpool--VmlldzoyMzM5Mjcw)

### Reference

* [DNA: Proximal Policy Optimization with a Dual Network Architecture](https://arxiv.org/abs/2206.10027)

cc @maitchison

## Types of changes
- [ ] Bug fix
- [ ] New feature
- [x] New algorithm
- [x] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x ] I have ensured `pre-commit run --all-files` passes (**required**).
- [x ] I have updated the documentation.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [ ] I have contacted [vwxyzjn](https://github.com/vwxyzjn) to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [x ] I have explained note-worthy implementation details.
    - [x ] I have explained the logged metrics.
    - [x ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [x ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [x ] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [x ] I have added links to the tracked experiments.
- [ ] I have updated the tests accordingly (if applicable).

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,474 @@
+        help="rollout horizon: the number of steps to run in each environment per policy rollout")
+    parser.add_argument("--anneal-lr", type=lambda x: bool(strtobool(x)), default=True, nargs="?", const=True,
+        help="Toggle learning rate annealing for policy and value networks")
+    parser.add_argument("--gamma", type=float, default=0.99,
```
Note a complete review yet, just putting a note here: the author suggests the `gamma` parameter used in DNA should be 0.999. I am ok with using either 0.99 as long as this difference is spelled out.

maitchison replied:
I'd recommend using the same settings as the paper, which would be the following changes to the defaults

gamma=0.999 
clip_coef=0.2
anneal-lr=false 
max_grad_norm=5.0

Gamma and clip_coef  make a fairly big difference too. The other settings (max_grad_norm, anneal etc) are not so important.

jseppanen replied:
yeah, I was originally thinking to have minimal changes from vanilla PPO hyperparams, but I agree it's better to have well tuned algorithm specific defaults out of the box.

jseppanen replied:
I'm also updating entropy coef, as it's 0.001 in the paper, vs. 0.01 in baseline PPO

maitchison commented:
Small thing  

> Here's the episodic rewards after 200M environment steps (50M gradient updates), compared to Fig. 6 in the original paper:

should be 

> Here's the episodic rewards after 200M environment steps (50M environment interactions), compared to Fig. 6 in the original paper:

The algorithm will make many more than 50M gradient updates due to the number of mini-batches.

maitchison commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,474 @@
+        num_threads=args.envpool_num_threads,
+        episodic_life=True,
+        reward_clip=True,
+        seed=args.seed,
```
Just a note. We designed DNA to run on the original rewards, and it works best with reward_clip off.  That being said, this would mean enabling reward normalization (i.e. dividing by the variance of returns).

jseppanen replied:
yeah I was actually thinking about the same thing, and maybe that could explain (part of) the differences in the atari-5 repro results.

jseppanen replied:
I removed the clipping and added reward normalization

maitchison commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,474 @@
+        agent_policy.zero_grad(True)  # don't clone gradients
+        old_agent_policy = deepcopy(agent_policy)
+        old_agent_policy.eval()
+        for epoch in range(args.distill_update_epochs):
```
Just a note, in future work we found by switching the order from Rollout-Value-Policy-Distil, to Rollout-Distil-Policy-Value we could avoid an additional forward (and the copy you need here too) and instead simply use the policy that was generated during rollout. This means the policy updates are slightly off-policy, even so, imperically it works slightly better than the order used in our paper.

jseppanen replied:
sounds great, I'll flip the order!

maitchison commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,474 @@
+import torch.optim as optim
+from torch.distributions import Categorical, kl_divergence
+from torch.utils.tensorboard import SummaryWriter
+
```
You'll probably need to add 

> torch.backends.cuda.matmul.allow_tf32 = False

before the first torch operation, otherwise, you may have issues with 30xx series cards. This issue has been fixed in torch 1.12, but I see we are pinned to an earlier version. Without it, some operations will be reduced to 10-bits of precision (around 3 sig fig) which may cause problems with, for example, the observation normalization. See (https://pytorch.org/docs/stable/notes/cuda.html#tf32-on-ampere) for a discussion on this issue.

maitchison commented:
@jseppanen This latest change looks great. It seems, however, that I made a typo in the paper. The entropy bonus should be 1e-2 (0.01) not 0.001 as I have mistakenly written in the paper. I think 1e-3 will likely cause the algorithm to underperform the default 0.01 setting. I've made a note of this and will fix it in the next revision. Thanks for spotting this!

jseppanen commented:
@maitchison OK I addressed all of your comments, plus added one extra change (used builtin `gym.wrappers.NormalizeObservation` instead of DIY) – I'm still going to re-run some of the experiments to verify that it still works 😅

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--exp-name", type=str, default=os.path.basename(__file__).rstrip(".py"),
+        help="the name of this experiment")
+    parser.add_argument("--seed", type=int, default=np.random.randint(2 ** 31),
```
Please revert the default seed to 1. This is an expected behavior of using CleanRL that the seeds by default is the same for reproducibility and debugging reason.

jseppanen replied:
👍

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+    # Algorithm specific arguments
+    parser.add_argument("--env-id", type=str, default="Pong-v5",
+        help="the id of the environment")
+    parser.add_argument("--total-timesteps", type=int, default=10_000_000,
```
Here we use the style `10000000`. I'd be open to using `10_000_000`, but this change should apply to all of our variants in the form of another PR.

 It's probably best to automate this using something like a pre-commit hook that automatically converts all instances of `10000000` to `10_000_000`.

jseppanen replied:
👍

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+        help="total timesteps of the experiments")
+    parser.add_argument("--num-envs", type=int, default=128,
+        help="the number of parallel game environments")
+    parser.add_argument("--envpool-num-threads", type=int, default=2,
```
This change should also apply to all envpool scripts in another PR.

jseppanen replied:
OK yeah, I'll drop it from here

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+    parser.add_argument("--envpool-num-threads", type=int, default=2,
+        help="the number of envpool worker threads")
+    parser.add_argument("--num-steps", type=int, default=128,
+        help="rollout horizon: the number of steps to run in each environment per policy rollout")
```
This "rollout horizon: " change should also apply to all PPO scripts in another PR.

jseppanen replied:
I think I'll just revert this

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+
+    parser.add_argument("--policy-learning-rate", type=float, default=2.5e-4,
+        help="the learning rate of the policy optimizer")
+    parser.add_argument("--policy-batch-size", type=int, default=2048,
```
put a note here

jseppanen replied:
I don't follow, you mean a comment explaining why there is an empty line? or something else?

vwxyzjn replied:
Sorry, that sentence was mainly for me. My original thought was that the other PPO related scripts have used `--num-minibatches`, so in this case we might be able to change it to `--num-policy-minibatches`. I will let you decide on this though. Feel free to pick one that feels like the cleanest design to you.

jseppanen replied:
Ah true; I changed the config to use minibatch size, to be able to easily match/verify the values against the the paper

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+            delta = rewards[t] + gamma * nextvalues * nextnonterminal - values[t]
+            advantages[t] = lastgaelam = delta + gamma * gae_lambda * nextnonterminal * lastgaelam
+        returns = advantages + values
+    else:
+        returns = torch.zeros_like(rewards)
+        for t in reversed(range(total_steps)):
+            if t == total_steps - 1:
+                nextnonterminal = 1.0 - next_done
+                next_return = next_value
+            else:
+                nextnonterminal = 1.0 - dones[t + 1]
+                next_return = returns[t + 1]
+            returns[t] = rewards[t] + gamma * nextnonterminal * next_return
+        advantages = returns - values
```
This can be removed. See https://github.com/vwxyzjn/cleanrl/issues/207

jseppanen replied:
but it looks like that code is still there in both `ppo_atari.py` and `ppo_atari_envpool.py` as well – maybe I'll keep it in for now, and then later all three files can be changed with a dedicated PR, wdyt?

vwxyzjn replied:
Either way works for me since #207 is a planned PR. If you decide to leave it in this PR, then #207 will take care of it :)

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+        self.critic = layer_init(nn.Linear(512, 1), std=1)
+
+    def forward(self, x):
+        """Run policy and value functions."""
```
This comment is not necessary.

jseppanen replied:
👍

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+        value = self.critic(hidden).squeeze(-1)
+        return logits, value
+
+    @staticmethod
+    def sample_action(logits):
+        probs = Categorical(logits=logits)
+        action = probs.sample()
+        return action, probs.log_prob(action)
+
+    @staticmethod
+    def action_log_prob(logits, action):
+        probs = Categorical(logits=logits)
+        return probs.log_prob(action), probs.entropy()
```
Please keep this part of the code as similar to the following as you can.

```
    def get_value(self, x):
        return self.critic(self.network(x / 255.0))

    def get_action_and_value(self, x, action=None):
        hidden = self.network(x / 255.0)
        logits = self.actor(hidden)
        probs = Categorical(logits=logits)
        if action is None:
            action = probs.sample()
        return action, probs.log_prob(action), probs.entropy(), self.critic(hidden)
```

To still make it work w/ the distillation loss, maybe you can do

`return action, probs.log_prob(action), probs.entropy(), self.critic(hidden), logits`

jseppanen replied:
done

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # Fix multiplication precision on Ampere devices (30xx) prior to PyTorch 1.12
+    # https://pytorch.org/docs/stable/notes/cuda.html#tf32-on-ampere
+    torch.backends.cuda.matmul.allow_tf32 = False
```
Could you do some preliminary experiments to check if this makes a difference? It's important that we understand the effect of the techniques added.

jseppanen replied:
unfortunately I don't have access to Ampere devices, so maybe I'll revert this as well

vwxyzjn replied:
I have ampere devices - once things are finalized I will be happy to give it a try.

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+    envs.single_action_space = envs.action_space
+    envs.single_observation_space = envs.observation_space
+    envs = RecordEpisodeStatistics(envs)
+    # if args.capture_video:
+    #     envs = gym.wrappers.RecordVideo(envs, f"videos/{run_name}")
```
Great catch. Please remove this.

jseppanen replied:
👍

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+    envs = RecordEpisodeStatistics(envs)
+    # if args.capture_video:
+    #     envs = gym.wrappers.RecordVideo(envs, f"videos/{run_name}")
+    envs = gym.wrappers.NormalizeObservation(envs)
+    envs = gym.wrappers.TransformObservation(envs, lambda obs: np.clip(obs, -3, 3))
```
I am a little hesitant to use `gym.wrappers.NormalizeObservation(envs)` in image-based envs. Dividing the pixels by 255 is a well-known normalization technique. `ppo_procgen.py`, for example, uses only the `gym.wrappers.NormalizeReward`.

https://github.com/vwxyzjn/cleanrl/blob/c41bee932ef998fcbbe301c6d91c2328c1c03ab7/cleanrl/ppo_procgen.py#L188-L197

To this end, my suggestion is not to use wrapper if it does not result in performance benefits.

jseppanen replied:
well this is meant to reproduce the DNA paper, where they suggest doing exactly that (normalize images to zero mean unity variance, then clip to -3...3)

should I add a comment to explain why, or no?

vwxyzjn replied:
If the paper specifically normalizes like this way then it's fine.

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+        return kl_divergence(old_probs, new_probs).mean()
+
+
+def main():
```
Out of curiosity, what is the purpose of using the `main()` function? We loss a bit of flexibility in debug when we do something like 

```
python ppo_dna_atari_envpool.py

# run it for 10 seconds then ctrl + c to print variables
```

jseppanen replied:
yeah, it's because I extracted `compute_advantages` as a separate function, but then noticed that it was using values of global variables defined later in the module, which is not what I intended. It's essentially a safeguard to remove all global variables with generic names like `reward` or `value` etc.

jseppanen replied:
Of course, then I fixed the compute_advantages arguments list to include all used variables, so it's no longer using global variables (I think), but in general, I feel it would be nicer that the program fails with an undefined variable instead of silently using a global variable defined somewhere else

However, I realize that this is not a problem if everything is in global scope and there are no functions (which pretty much is the cleanRL style), but I think that the compute_advantages function is necessary in this case, and that's why I opted for increased safety against (future) bugs, at the cost of decreased interactive convenience

Personally, I edit the file locally to replace `def main():` with `if 1:` for a smooth interactive experience, if needed. But I'm happy to revert this change as well if you think so; lmk!

Dipamc77 replied:
If we'd like to have all the variables globally, its possible to add two extra lines in the original GAE calculation code to have the two separate advantage calculation steps. I think almost everything else remains same.

```
delta_policy = rewards[t] + gamma * nextvalues * nextnonterminal - values[t]
advantages_policy[t] = lastgaelam = delta_policy + gamma * policy_lambda * nextnonterminal * lastgaelam
delta_value = rewards[t] + gamma * nextvalues * nextnonterminal - values[t]
advantages_value[t] = lastgaelam = delta_value + gamma * value_lambda * nextnonterminal * lastgaelam
``` 

Generally I prefer functions more but cleanrl avoids them so this can be an alternative.

vwxyzjn replied:
Thanks @Dipamc77 for the suggestion. In this case though, using `compute_advantages` seems pretty clean to me, so I am ok with it :)

vwxyzjn replied:
>  It's essentially a safeguard to remove all global variables with generic names like reward or value etc.

This is a good point. I was thinking of using a pre-commit hook like pylint to detect this kind of situations. For now, let's revert back to not using the `main()` function to be consistent w/ other scripts

jseppanen replied:
done

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+    value_optimizer = optim.Adam(agent_value.parameters(), lr=args.value_learning_rate, eps=1e-5)
+    distill_optimizer = optim.Adam(agent_policy.parameters(), lr=args.distill_learning_rate, eps=1e-5)
+
+    if args.track:
+        # log gradients
+        wandb.watch(agent_policy)
+        wandb.watch(agent_value)
```
Is gradient logging useful for debugging?

jseppanen replied:
yeah it's sometimes been useful, esp. when trying to choose the gradient clipping value, but I can remove this to be in line with the baseline PPO style

vwxyzjn replied:
Interesting - could you elaborate on the process of choosing the gradient clipping value?

jseppanen replied:
well I just eyeball the gradient norms over the course of training and compare against returns at the same time, and try to guess if SGD updates might be overshooting

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    env_step = 0
```
Not necessary. We know the frame skip number already, and we can post-process this.

jseppanen replied:
👍

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+
+    for update in range(1, num_updates + 1):
+        # Annealing the rate if instructed to do so.
+        if args.anneal_lr:
+            frac = 1.0 - (update - 1.0) / num_updates
+            policy_optimizer.param_groups[0]["lr"] = frac * args.policy_learning_rate
+            value_optimizer.param_groups[0]["lr"] = frac * args.value_learning_rate
+            distill_optimizer.param_groups[0]["lr"] = frac * args.distill_learning_rate
+
+        for step in range(0, args.num_steps):
+            global_step += 1 * args.num_envs
+            env_step += envs.spec.config.frame_skip * args.num_envs
+            obs[step] = next_obs
+            dones[step] = next_done
+
+            # ALGO LOGIC: action logic
+            with torch.no_grad():
+                logit, _ = agent_policy(next_obs)
+                _, value = agent_value(next_obs)
+                action, logprob = Agent.sample_action(logit)
+            action_logits[step] = logit
+            actions[step] = action
+            logprobs[step] = logprob
+            values[step] = value
+
+            # TRY NOT TO MODIFY: execute the game and log data.
+            next_obs, reward, done, info = envs.step(action.cpu().numpy())
+            rewards[step] = torch.tensor(reward, dtype=torch.float32).to(device).view(-1)
+            next_obs = torch.tensor(next_obs, dtype=torch.float32).to(device)
+            next_done = torch.tensor(done, dtype=torch.float32).to(device)
```
This difference should be kept to a minimum.

<img width="1469" alt="image" src="https://user-images.githubusercontent.com/5555347/180087529-2a001df2-92c0-485f-969b-90f20c5ca01e.png">

jseppanen replied:
done

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,468 @@
+    envs.close()
+    writer.close()
+
+    if args.track:
+        # save final model
+        torch.save(agent_policy.state_dict(), "agent_policy.pt")
+        torch.save(agent_value.state_dict(), "agent_value.pt")
+        wandb.save("agent_policy.pt")
+        wandb.save("agent_value.pt")
```
Out of curiosity, what is the use case of this?

jseppanen replied:
removed

vwxyzjn replied:
How would you generate gameplay videos, though? Through the gym wrappers? This might be a legitimate use case for model saving.

vwxyzjn reviewed (changes requested):
Thank you so much for your contribution. The code has great readability, and thank you for including the documentation and benchmark results.

Most of my comments are related to style - generally speaking, we try to minimize the lines of different code across files. Such a practice has two benefits
1. the filediff clearly shows the necessary changes for implementing another algorithm,
2. it makes our single-file codebase more maintainable.

To this end, please try to keep the original structure as much as possible, and I suggest walking through the code difference between `ppo_atari_envpool.py` and `ppo_dna_atari_envpool.py` and remove any unnecessary changes.


<img width="1483" alt="image" src="https://user-images.githubusercontent.com/5555347/180083940-216f21a9-8e1f-4faa-9165-1274d8c9f672.png">

jseppanen commented:
OK so some news:

I reverted the hyperparameter change (commit 0fe7b1f) and the learning stage ordering change (commit e50bf9e). The reason is that I did some investigation to what changes led to the algorithm not being able to solve Pong in 3M steps. What I found out was that:
1. using gym observation normalization (commit cde9ada) and removing reward clipping & adding reward normalization (commit 7bd5244) alone were good, no performance regressions there
2. however, when I applied the hyperparameter change (commit 0fe7b1f) on top of observation & reward normalization, then it could suddenly no longer learn Pong
3. to cross-check, I also tested only applying the new hyperparameters without any other changes from original version (commit 981201f), and although it was able to start learning Pong, there was a reduction in learning speed/sample efficiency
I didn't want to start digging into which specific hyperparameter was triggering the problems but simply reverted the whole change for now.

Then I had also noticed another regression in wall clock learning speed, and also, because @vwxyzjn encouraged me to try and keep the code changes as small as possible, compared to the baseline `ppo_atari_envpool.py` algorithm, I also decided to revert the learning stage order change (commit e50bf9e). Maybe that could be a separate PR one day, with separate experiments to back it up.

Finally, I went through the whole file and reverted unnecessary differences compared to `ppo_atari_envpool.py` and this brings us to the current commit, 94fc331. Based on quick Pong experiment, it's working as well as the original version, so I'll be launching hopefully final experiments to make sure that it's still working like the original implementation.

The hyperparameters are a bit of a mystery, and it would be nice to get more visibility on what's going on there, but I'm afraid that would require some hyperparameters sweeping on a good set of Atari games, and I'm not sure I'm in the mood for that right now 😅

Dipamc77 commented:
Code looks good, matches details mentioned in the paper. Really like the idea that just using a different lambda for advantage estimation gives such a huge boost overall.

One important difference seems that PPO-DNA consistently uses 128 parallel envs and PPO by default is using 8. I'm not aware if using higher makes too much of a difference in PPO but selecting 128 for PPO-DNA seems to be a conscious choice (paper does give some explanation on this as well)

Side note, the paper cites PPG as the closest work but I'd argue this is closer [Decoupling Value and Policy for Generalization in Reinforcement Learning](https://arxiv.org/pdf/2102.10330.pdf)

vwxyzjn commented:
> PPO-DNA consistently uses 128 parallel envs

This kind of changes things a lot. I am more inclined to reproduce this work similar to how the paper is done. Please try to use 128 parallel envs if possible.

jseppanen commented:
Update: I re-ran the experiments with commit 94fc331 and updated the documentation/results, and removed the main() function.

@vwxyzjn github is telling me that "requested changes must be addressed" but I can't find which changes are still needed, sorry for that. I think I addressed all your comments but is there something still that I missed?

jseppanen commented:
> One important difference seems that PPO-DNA consistently uses 128 parallel envs and PPO by default is using 8. I'm not aware if using higher makes too much of a difference in PPO but selecting 128 for PPO-DNA seems to be a conscious choice (paper does give some explanation on this as well)

Yeah I noticed a big improvement when using 128 envs, so all the experiments were done with that.

maitchison commented:
> 

The large number of paralell agents helps a lot with dealing the gradient noise. Parallel agents are better than longer n_steps due to the n_steps producing temporally correlated trajectories. One of the big tricks to getting PG to work well is using *a lot* of parallel agents. From memory the Hide-and-seek paper (https://arxiv.org/abs/1909.07528) used batchs of tens-of-thousands. ProcGen uses 64k. I've tried higher than 128, but gains diminish quickly after that on Atari, and memory becomes an issue. 

A large number of Parallel is also compute efficient as forwarding the model and parallelizing the environments all become faster.

maitchison commented:
> Code looks good, matches details mentioned in the paper. Really like the idea that just using a different lambda for advantage estimation gives such a huge boost overall.
> 
> One important difference seems that PPO-DNA consistently uses 128 parallel envs and PPO by default is using 8. I'm not aware if using higher makes too much of a difference in PPO but selecting 128 for PPO-DNA seems to be a conscious choice (paper does give some explanation on this as well)
> 
> Side note, the paper cites PPG as the closest work but I'd argue this is closer [Decoupling Value and Policy for Generalization in Reinforcement Learning](https://arxiv.org/pdf/2102.10330.pdf)

Good spotting! We are actually in the process now of updating the paper to discuss this reference :)

maitchison commented:
> OK so some news:
> 
> I reverted the hyperparameter change (commit [0fe7b1f](https://github.com/vwxyzjn/cleanrl/commit/0fe7b1f1b4d846cc6e61e5ddb1547e5664bbb6e3)) and the learning stage ordering change (commit [e50bf9e](https://github.com/vwxyzjn/cleanrl/commit/e50bf9e220360e3fc049316c1dbaa825f15b9e4b)). The reason is that I did some investigation to what changes led to the algorithm not being able to solve Pong in 3M steps. What I found out was that:
> 
> 1. using gym observation normalization (commit [cde9ada](https://github.com/vwxyzjn/cleanrl/commit/cde9adaf46629bb406e1c15c84ff565fcb5cdf6f)) and removing reward clipping & adding reward normalization (commit [7bd5244](https://github.com/vwxyzjn/cleanrl/commit/7bd52441e561168b7fdfe1f089e61f2a90144dc1)) alone were good, no performance regressions there
> 2. however, when I applied the hyperparameter change (commits [0fe7b1f](https://github.com/vwxyzjn/cleanrl/commit/0fe7b1f1b4d846cc6e61e5ddb1547e5664bbb6e3) & [1b91a46](https://github.com/vwxyzjn/cleanrl/commit/1b91a46a74f49c2507de31e931f375fad4b1cb88)) on top of observation & reward normalization, then it could suddenly no longer learn Pong
> 3. to cross-check, I also tested only applying the new hyperparameters without any other changes from original version (commit [981201f](https://github.com/vwxyzjn/cleanrl/commit/981201f39978c7162a437a256a616cd03ff1aaf4)), and although it was able to start learning Pong, there was a reduction in learning speed/sample efficiency
> 
> I didn't want to start digging into which specific hyperparameter was triggering the problems but simply reverted the whole change for now.
> 
> Then I had also noticed another regression in wall clock learning speed, and also, because @vwxyzjn encouraged me to try and keep the code changes as small as possible, compared to the baseline `ppo_atari_envpool.py` algorithm, I also decided to revert the learning stage order change (commit [e50bf9e](https://github.com/vwxyzjn/cleanrl/commit/e50bf9e220360e3fc049316c1dbaa825f15b9e4b)). Maybe that could be a separate PR one day, with separate experiments to back it up.
> 
> Finally, I went through the whole file and reverted unnecessary differences compared to `ppo_atari_envpool.py` and this brings us to the current commit, [94fc331](https://github.com/vwxyzjn/cleanrl/commit/94fc331d2ecdcbe70279e72d719787a3e75f83f1). Based on quick Pong experiment, it's working as well as the original version, so I'll be launching hopefully final experiments to make sure that it's still working like the original implementation.
> 
> The hyperparameters are a bit of a mystery, and it would be nice to get more visibility on what's going on there, but I'm afraid that would require some hyperparameters sweeping on a good set of Atari games, and I'm not sure I'm in the mood for that right now sweat_smile

I'm not sure if this is helpful, but I was reading through some of my old notes and found a note saying DNA did not converge on Pong without observation normalization. Perhaps the regression could have caused issues by the normalization? We normed to unit normal (via a running mean and second moment), then clipped to -5,+5, then multiplied by 3.0. But anything reasonable should do. We also used the default pytorch weight initialization, rather than orthogonal. I don't think that'd matter much though.

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,432 @@
+    device = torch.device("cuda" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = envpool.make(
```
Did you decide not to use sticky action? See `repeat_action_probability` in envpool's documentation. Also CC @maitchison .

jseppanen replied:
I took the config from `ppo_atari_envpool.py` which doesn't use sticky actions, and I wouldn't like to change that now, because that would unfortunately require re-running all experiments.

vwxyzjn commented on docs/rl-algorithms/ppo_dna.md:
```diff
@@ -0,0 +1,101 @@
+
+| Environment      | `ppo_dna_atari_envpool.py` | `ppo_atari_envpool.py` |
+| ----------- | ----------- | ----------- | 
+| BattleZone-v5 (40M steps) | 74000 ± 15300 | 28700 ± 6300 |
```
Do you mean 50M steps?

jseppanen replied:
Actually, my baseline `ppo_atari_envpool.py` experiments crashed before reaching full 50M steps, but I think the results are still valuable because they quite clearly show the performance difference between CleanRL baseline PPO and PPO-DNA.

vwxyzjn commented on docs/rl-algorithms/ppo_dna.md:
```diff
@@ -0,0 +1,101 @@
+| ----------- | ----------- | ----------- | 
+| BattleZone-v5 (40M steps) | 74000 ± 15300 | 28700 ± 6300 |
+| BeamRider-v5 (10M steps) | 5200 ± 900 | 1900 ± 530 |
+| Breakout-v5 (10M steps) | 319 ± 63 | 349 ± 42 |
```
I suggest being consistent and always using 50M steps. It's ok to compare the `ppo_atari_envpool.py` that only has 10M steps. If you think comparing 50M steps is warranted, please start a `ppo_atari_envpool_50M` wandb project in the `openrlbenchmark` entity and run `ppo_atari_envpool.py` in 50M steps there.

jseppanen replied:
Sure, though I'm using rented cloud GPU's so re-running has a direct cost to me.

Also, I think there's already a couple of layers of evidence for PPO-DNA without running any more experiments:
1. Primarily, I was interested in reproducing the paper results on the five Atari envs shown in Fig 6 in the paper. The paper used 50M steps for their experiments, so that's what I did as well. My results are not exactly the same, but in the same ballpark (one env performs better, two are comparable, and two are worse).
2. Secondarily, I wanted to do an apples-to-apples comparison against the baseline `ppo_atari_envpool.py` implementation, because that way the environment settings and other implementation details wouldn't distort the results. The CleanRL documentation has learning curves for Breakout, Pong, and BeamRider for 10M steps, so that's what I ran (plus I noticed some Tennis runs in the W&B project). I attempted to run the comparisons for 50M steps as well, but the baseline runs crashed (due to my cloud account running dry), so I thought 40M steps is already more than enough.

Therefore, I'm not sure what the benefit from re-running `ppo_atari_envpool.py` for 50M steps would be?

With regards to hyperparameters, they are otherwise the same, except for total_timesteps.

vwxyzjn commented on docs/rl-algorithms/ppo_dna.md:
```diff
@@ -0,0 +1,101 @@
+| DoubleDunk-v5 (40M steps) | -4.1 ± 1.0 | -2.0 ± 0.8 |
+| NameThisGame-v5 (40M steps) | 19100 ± 2300 | 4400 ± 1200 |
+| Phoenix-v5 (45M steps) | 186000 ± 67000 | 9900 ± 2700 |
+| Pong-v5 (3M steps) | 19.5 ± 1.0 | 16.6 ± 2.4 |
```
Same comment as above. Please run using 50M steps.

jseppanen replied:
What I could do instead is maybe show the results at the 10M steps mark, if 40M is a problem?

vwxyzjn reviewed (commented):
Thanks for your patience! You have addressed my comments and please ignore the `requested changes` from Github. 

The next step is to contact me to have yourself added in the `openrlbenchmark` wandb team. Then you may either move your experiments in `openrlbenchmark/cleanrl` or re-run some of them.

I added some small refactoring in the last two commits. Besides, I updated the checklist to include a few more items. 

- [ ]  I have updated the overview sections at the [docs](https://docs.cleanrl.dev/rl-algorithms/overview/) and the [repo](https://github.com/vwxyzjn/cleanrl#overview)
- [ ] I have added the commands used to run experiments in the `benchmark` folder like [benchmark/ppo.sh](https://github.com/vwxyzjn/cleanrl/blob/master/benchmark/ppo.sh)

Please also add an end-to-end test in 

https://github.com/vwxyzjn/cleanrl/blob/df3ab2a1843b906c6e528d37e1eb0a34e831e5db/tests/test_envpool.py#L4-L9

vwxyzjn commented:
Starting a new thread for discussion. @jseppanen, thanks for bearing with me. The results are definitely convincing, but the main issue is consistency (same configuration, same steps, non-crashed runs). Totally understand that running experiments can be expensive. I can help run the experiments, but it will take some time (probably a month).

jseppanen commented:
@vwxyzjn sure, I added benchmarks/ppo_dna.sh.  Also, maybe I haven't communicated my experiment results clearly enough, I could also try and consolidate them into one place.

bragajj commented:
@jseppanen wondering if you're still interested in this PR. I took a quick look at your wandb and see runs in your account for Phoenix-v5, NamesThisGame-v5, DoubleDunk-v5 at 50M steps, just not Pong. Trying to understand where we ended up here.

Was it that you were reporting results at different step counts despite having 50M step runs completed? Or were those runs not completed successfully? Just want to see if I can provide any help or guidance.

maitchison commented:
Hi @vwxyzjn, I now have some time I can put into this and would be happy to finish off the last few things that need doing. Looks like @jseppanen has got it mostly there, so it shouldn't take too long. I have access to a cluster I can run any additional experiments if needed.

vwxyzjn commented:
Hey, sorry folks for not replying sooner. I looked into the PR a bit more and looks like @jseppanen has already addressed my comments on 50M steps. Thanks a lot. I generated the following plots using https://github.com/vwxyzjn/ppo-atari-metrics/blob/main/rlops.py


```
python rlops.py --wandb-project-name envpool-atari \
    --wandb-entity openrlbenchmark \
    --filters 'ppo_dna_atari_envpool_94fc331?wpn=cleanrl&we=jseppanen' 'ppo_atari_envpool_xla_jax_truncation?metric=charts/avg_episodic_return'   \
    --env-ids BattleZone-v5 DoubleDunk-v5 NameThisGame-v5 Phoenix-v5 Qbert-v5 Pong-v5 BeamRider-v5 Breakout-v5 Tennis-v5 \
    --output-filename compare.png --scan-history
```

<img width="1475" alt="image" src="https://user-images.githubusercontent.com/5555347/202879349-0899bb78-0a49-4fde-a2a9-a7b665e83f79.png">

The results look good. **I will go ahead and run the Pong-v5 experiments to match the 10M steps, and that should be all for all of the experiments.** @jseppanen would you mind moving the runs from your entity to `openrlbenchmark`? You can move them as shown in the video below:

https://user-images.githubusercontent.com/5555347/202879526-c2b54798-a103-4192-a1f1-b3183abc176f.mov



# A note on environment preprocessing

The preprocessing steps should be like below, but `full_action_space=True` is not currently supported by EnvPool (https://github.com/sail-sg/envpool/issues/220). Let's put this note in the documentation and not block this PR any longer.

```python
envs = envpool.make(
    args.env_id,
    env_type="gym",
    num_envs=args.num_envs,
    episodic_life=False,  # Machado et al. 2017 (Revisitng ALE: Eval protocals) p. 6
    repeat_action_probability=0.25,  # Machado et al. 2017 (Revisitng ALE: Eval protocals) p. 12
    noop_max=1,  # Machado et al. 2017 (Revisitng ALE: Eval protocals) p. 12 (no-op is deprecated in favor of sticky action)
    max_episode_steps=int(108000 / 4),  # Hessel et al. 2018 (Rainbow DQN), Table 3, Max frames per episode
    reward_clip=True,
    seed=args.seed,
    # full_action_space=True, # currently not supported by EnvPool Machado et al. 2017 (Revisitng ALE: Eval protocals) Table 5
)
```

vwxyzjn commented on cleanrl/ppo_dna_atari_envpool.py:
```diff
@@ -0,0 +1,426 @@
+                end = start + args.value_batch_size
+                mb_inds = b_inds[start:end]
+
+                newvalue = agent_value.get_value(b_obs[mb_inds])
+
```
I hit a major regression with the latest script. It appears it's because `newvalue` here is not flattened

<img width="1362" alt="image" src="https://user-images.githubusercontent.com/5555347/203342912-5d770322-b656-4792-bc09-0fd6986e97f9.png">
